stackslib/src/net/download/nakamoto.rs

```diff
@@ -0,0 +1,4283 @@
+                continue;
+            }
+            if !last_available_tenures.contains(&downloader.tenure_end_block_id) {
+                continue;
+            }
+            test_debug!(
+                "Transition downloader for {} from waiting to fetching",
+                &downloader.tenure_id_consensus_hash
+            );
+            if let Err(e) = downloader.transition_to_fetch_end_block() {
+                warn!(
+                    "Downloader for {} failed to transition to fetch end block: {:?}",
+                    &downloader.tenure_id_consensus_hash, &e
+                );
+            }
+        }
+    }
+
+    /// Create a given number of downloads from a schedule and availability set.
+    /// Removes items from the schedule, and neighbors from the availability set.
+    /// A neighbor will be issued at most one request.
+    pub(crate) fn make_tenure_downloaders(
+        &mut self,
+        schedule: &mut VecDeque<ConsensusHash>,
+        available: &mut HashMap<ConsensusHash, Vec<NeighborAddress>>,
+        tenure_block_ids: &HashMap<NeighborAddress, AvailableTenures>,
+        count: usize,
+        agg_public_keys: &BTreeMap<u64, Option<Point>>,
+    ) {
+        test_debug!("schedule: {:?}", schedule);
+        test_debug!("available: {:?}", &available);
+        test_debug!("tenure_block_ids: {:?}", &tenure_block_ids);
+        test_debug!("inflight: {}", self.inflight());
+        test_debug!(
+            "count: {}, running: {}, scheduled: {}",
+            count,
+            self.num_downloaders(),
+            self.num_scheduled_downloaders()
+        );
+
+        self.clear_available_peers();
+        self.clear_finished_downloaders();
+        self.try_transition_fetch_tenure_end_blocks(tenure_block_ids);
+        while self.inflight() < count {
+            let Some(ch) = schedule.front() else {
+                break;
+            };
+            if self.completed_tenures.contains(&ch) {
+                test_debug!("Already successfully downloaded tenure {}", &ch);
+                schedule.pop_front();
+                continue;
+            }
+            let Some(neighbors) = available.get_mut(ch) else {
+                // not found on any neighbors, so stop trying this tenure
+                test_debug!("No neighbors have tenure {}", ch);
+                schedule.pop_front();
+                continue;
+            };
+            if neighbors.len() == 0 {
+                // no more neighbors to try
+                test_debug!("No more neighbors can serve tenure {}", ch);
+                schedule.pop_front();
+                continue;
+            }
+            let Some(naddr) = neighbors.pop() else {
+                test_debug!("No more neighbors can serve tenure {}", ch);
+                schedule.pop_front();
+                continue;
+            };
+            if self.try_resume_peer(naddr.clone()) {
+                continue;
+            };
```

Comment:
What's happening in this case?

It seems like `naddr` is either: already working on this tenure, already working on another downloader, now gets assigned to a different downloader, or now gets assigned to this `ch`. In any of those cases, the loop then iterates. When the loop iterates, `ch` is still at the front of the queue, so we come back here and do the same thing with a different peer. That is unless we've run out of peers for this `ch`, in which case we pop `ch`, but it seemingly had nothing to do with whether or not a peer was assigned to download for the tenure or even if there is a downloader for the tenure at all. So, I'm not really following the logic of this loop.

Reply:
I gave an example here: https://github.com/stacks-network/stacks-core/pull/4476#discussion_r1531003266.  The gist of it is, you (1) only want one in-flight HTTP request per peer, and (2) sometimes, there are fewer peers than downloader state machines.  So, on each pass, we need to find downloaders that still have some progress to make but currently have no in-flight messages, and assign them a peer to which to send their next request.

Reply:
And yes, the loop does iterate in order to drain the `schedule` (or until we've created as many downloaders as we've been configured to create, whichever happens first).  The `ch` gets popped if it has a downloader already, so the next tenure can be given a downloader and a peer to service HTTP requests.